inspirehep/modules/workflows/tasks/refextract.py

```diff
@@ -191,42 +191,83 @@ def extract_references_from_raw_ref(reference, custom_kbs_file=None):
-
-    config_default = current_app.config['WORKFLOWS_REFERENCE_MATCHER_DEFAULT_CONFIG']
-    config_jcap_and_jhep = current_app.config['WORKFLOWS_REFERENCE_MATCHER_JHEP_AND_JCAP_CONFIG']
-    config_data = current_app.config['WORKFLOWS_REFERENCE_MATCHER_DATA_CONFIG']
+    config_unique_identifiers = current_app.config[
+        'WORKFLOWS_REFERENCE_MATCHER_UNIQUE_IDENTIFIERS_CONFIG']
+    config_default_publication_info = current_app.config[
+        'WORKFLOWS_REFERENCE_MATCHER_DEFAULT_PUBLICATION_INFO_CONFIG']
+    config_jcap_and_jhep_publication_info = current_app.config[
+        'WORKFLOWS_REFERENCE_MATCHER_JHEP_AND_JCAP_PUBLICATION_INFO_CONFIG']
+    config_data = current_app.config[
+        'WORKFLOWS_REFERENCE_MATCHER_DATA_CONFIG']
 
     journal_title = get_value(reference, 'reference.publication_info.journal_title')
-    if journal_title in ['JHEP', 'JCAP']:
-        config = config_jcap_and_jhep
-    else:
-        config = config_default
+    config_publication_info = config_jcap_and_jhep_publication_info if \
+        journal_title in ['JCAP', 'JHEP'] else config_default_publication_info
+
+    configs = [config_unique_identifiers, config_publication_info, config_data]
+
+    matched_reference, matched_recid = \
+        match_reference_with_config(reference, configs, previous_matched_recid)
 
-    result = next(match(reference, config), None)
-    if result:
-        matched_recid = result['_source']['control_number']
-        reference['record'] = get_record_ref(matched_recid, 'literature')
+    return reference, matched_recid
 
-    result_for_data = next(match(reference, config_data), None)
-    if result_for_data:
-        matched_recid = result_for_data['_source']['control_number']
-        reference['record'] = get_record_ref(matched_recid, 'data')
+
+def match_reference_with_config(reference, configs, previous_matched_recid=None):
+    """Match a reference using inspire-matcher given the config.
+
+    Args:
+        reference (dict): the metadata of the reference.
+        config (dict): the list of inspire-matcher configurations for queries.
+        previous_matched_recid (int): the record id of the last matched
+            reference from the list of references.
+
+    Returns:
+        dict: the matched references.
+    """
+    # XXX: avoid this type casting.
+    if get_value(reference, 'reference.publication_info.year'):
+        reference['reference']['publication_info']['year'] = str(
+            reference['reference']['publication_info']['year'])
+
+    matched_recid = None
+    for config in configs:
+        matched_records = dedupe_list(list(match(reference, config)))
+        for matched_record in matched_records:
+            matched_recid = matched_record['_source']['control_number']
+            if (len(matched_records) == 1) or (matched_recid == previous_matched_recid):
+                if get_value(config, 'index') == 'records-data':
```

Comment:
this function should return a recid only, so that the substistution for the record ref can be done by the caller.